### DISCUSSION TITLE
Bad certificate error code:0x8041012a

### DISCUSSION THREAD

**Discussion by shashenli** [microsoft/msquic#3048] [Q&A]
hello i need some help would u pls help me? thank u. i use self signed certificate on server and i have already added the CA root certificate in Windows certMgr's trusted root signing certificate but still 0x8041012a i dont know why

**Comment by shashenli** [comment]
and thats how i do everything details 
1.use openssl to make CA certificate privkey and server certificate privkey 
(1)openssl genrsa -out rootCA.key 2048
(2)openssl req -x509 -new -nodes -key rootCA.key -subj "/CN=*.tunnel.tonybai.com" -days 5000 -out rootCA.pem
(3)openssl genrsa -out device.key 2048
(4)openssl req -new -key device.key -subj "/CN=*.tunnel.tonybai.com" -out device.csr
(5)openssl x509 -req -in device.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out device.crt -days 5000
2.added CA certificate in Windows certMgr's trusted root signing certificate 
3.use quicsample to test if it worked
server parameters:-server -cert_file:./key/device.crt -key_file:./key/device.key
client parameters:-client -target:127.0.0.1
error code:Shut down by transport, 0x8041012a

**Reply by nibanks** [reply]
Did you put the certificate in the server's Root CA certificate store or the client's? If needs to be installed where the validation is performed (on the client). If it is on the client, the easiest way I know to manually validate the server's certificate is to simply double click the file on Windows to open up the certificate manager for it. Go through the details to see if it shows everything as valid.

**Reply by shashenli** [reply]
i put the cert in both client and server, u mean that if my cert is working fine,client on windows will go certificate manager to load cert and verify it?

**Reply by shashenli** [reply]
verify Server's cert sigend by my own CA

**Reply by shashenli** [reply]
im sure that my cert has no problem i guess that msquic client on windows just will not load cert which installed in OS.

**Reply by shashenli** [reply]
maybe this is a bug? im not sure. hope u could answer me thank u

> Did you put the certificate in the server's Root CA certificate store or the client's? If needs to be installed where the validation is performed (on the client). If it is on the client, the easiest way I know to manually validate the server's certificate is to simply double click the file on Windows to open up the certificate manager for it. Go through the details to see if it shows everything as valid.

**Reply by nibanks** [reply]
Can you post a picture when you open the certificate in Windows, showing it is completely valid? I'm not sure what problem you're having exactly. Our team isn't certificate experts.

**Reply by shashenli** [reply]
![cert](https://user-images.githubusercontent.com/93954749/189166554-03a6c2d3-bdeb-4782-b13f-701c7168423a.png)
![certMgr](https://user-images.githubusercontent.com/93954749/189166457-b09f0a07-e349-4d55-9d01-3e915d73bd1e.png)
i from china so thats all Chinese but u could still see that my cert is working fine


> Can you post a picture when you open the certificate in Windows, showing it is completely valid? I'm not sure what problem you're having exactly. Our team isn't certificate experts.

**Reply by nibanks** [reply]
@anrossi any ideas

**Reply by shashenli** [reply]
and in another q&a the person asking the question  also said that client on windows will not load cert which installed in OS
https://github.com/microsoft/msquic/discussions/2991#discussioncomment-3588522
